chemprop/cli/train.py

```diff
@@ -865,7 +866,21 @@ def train_model(
         trainer.fit(model, train_loader, val_loader)
 
         if test_loader is not None:
-            predss = trainer.predict(dataloaders=test_loader)
+            if isinstance(trainer.strategy, DDPStrategy):
+                torch.distributed.destroy_process_group()
+
+                best_ckpt_path = trainer.checkpoint_callback.best_model_path
+                trainer = pl.Trainer(
+                    logger=trainer_logger,
+                    enable_progress_bar=True,
+                    accelerator=args.accelerator,
+                    devices=1,
+                )
+                model = build_model(args, train_loader.dataset, output_transform, input_transforms)
+                model = model.load_from_checkpoint(best_ckpt_path)
+                predss = trainer.predict(model, dataloaders=test_loader)
```

Comment:
Your understanding is correct. DDP uses a distributed sampler, which drops samples to ensure the number of batches divides evenly across the GPUs. This is acceptable during validation to measure model performance. However, for testing, we should evaluate every data point. Even if the data points are evenly distributed across different processes, we would need to save the predictions from different processes, merge them, and also find the indices from the sampler. IMO, the inference of the D-MPNN model is not expensive, so I think using a single GPU here is fine.

Reply:
I think lightning does the opposite of what you are describing - look at the note at the bottom of this section of their docs: https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#testing

It says that samples will actually be duplicated when batches don't evenly divide across GPUs. It even suggests running `validate` (and I think it means to suggest for `test`, too) on only a single device to avoid this.

Reply:
Regardless of this I think what we have here is sound. My above comment might be a good thing to think about in the future, but I don't think it's a big deal.